crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -112,4 +167,23 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
             }
         }
     }
+
+    fn add_match_pattern_singleton(
+        &mut self,
+        subject: &ast::Expr,
+        pattern: &ast::PatternMatchSingleton,
+    ) {
+        if let Some(ast::ExprName { id, .. }) = subject.as_name_expr() {
+            // SAFETY: we should always have a symbol for every Name node.
+            let symbol = self.symbols().symbol_id_by_name(id).unwrap();
+
+            let ty = match pattern.value {
+                ast::Singleton::None => Type::None,
+                ast::Singleton::True => Type::BooleanLiteral(true),
+                ast::Singleton::False => Type::BooleanLiteral(false),
+            };
+            let constraint = IntersectionBuilder::new(self.db).add_positive(ty).build();
+            self.constraints.insert(symbol, constraint);
```

Comment:
An intersection with only one positive value in it always simplifies back to just that value, so we don't need to build an intersection at all here, it's a no-op. Given that intersection builders allocate, this might help with the regression.
```suggestion
            self.constraints.insert(symbol, ty);
```